server/accounts.go

```diff
@@ -1355,43 +1355,47 @@ func (a *Account) subscribeInternal(subject string, cb msgHandler) (*subscriptio
 		return nil, fmt.Errorf("no internal account client")
 	}
 
-	sub, err := c.processSub([]byte(subject+" "+sid), false)
+	sub, err := c.processSub(createSub(c, subject, sid, cb), false)
 	if err != nil {
 		return nil, err
 	}
 
-	sub.icb = cb
 	return sub, nil
 }
 
 // This will add an account subscription that matches the "from" from a service import entry.
 func (a *Account) addServiceImportSub(si *serviceImport) error {
 	a.mu.Lock()
 	c := a.internalClient()
-	sid := strconv.FormatUint(a.isid+1, 10)
-	a.mu.Unlock()
-
 	// This will happen in parsing when the account has not been properly setup.
 	if c == nil {
+		a.mu.Unlock()
 		return nil
 	}
-
 	if si.sub != nil {
+		a.mu.Unlock()
 		return fmt.Errorf("duplicate call to create subscription for service import")
 	}
+	a.isid++
+	sid := strconv.FormatUint(a.isid, 10)
+	cb := func(sub *subscription, c *client, subject, reply string, msg []byte) {
+		c.processServiceImport(si, a, msg)
+	}
+	si.sub = createSub(c, si.from, sid, cb)
+	sub := si.sub
+	a.mu.Unlock()
 
-	sub, err := c.processSub([]byte(si.from+" "+sid), true)
+	rsub, err := c.processSub(sub, true)
 	if err != nil {
 		return err
 	}
-
-	sub.icb = func(sub *subscription, c *client, subject, reply string, msg []byte) {
-		c.processServiceImport(si, a, msg)
-	}
-
 	a.mu.Lock()
-	a.isid++
-	si.sub = sub
+	// TODO: Normally, we don't have to set si.sub here since it was done prior to
```

Comment:
@derekcollison Actually, since we would fail above if we detect that si.sub is not nil, this is likely a situation that will never occur, so I could remove this code. About err checking, see comment in processSub().

Reply:
I have removed this part in the 2nd commit.

Reply:
ok